### DISCUSSION TITLE
Using with a base monad other that IO

### DISCUSSION THREAD

**Discussion by newhoggy** [haskell-effectful/effectful#272] [Q&A] [answered]
I'm interested in using `effectful` with a base monad other than `IO` - in particular `PropertyT IO` from the `hedgehog` library.

Is this possible?

**Comment by ocharles** [comment, marked as answer]
This isn't possible, but `PropertyT` is quite a strange monad that might not even possible to capture as an effect with `effectful`. What I have had success with is defining a `Test` effect that models Hedgehog's `TestT` monad, along with `evalEff`. This lets you combine effectful and `MonadTest`, which has felt like the goth balance. If this is interesting let me know, as I've been meaning to tidy this up and open source it

**Comment by newhoggy** [comment]
Yes, please.

**Comment by arybczak** [comment]
Btw, doesn't `PropertyT (Eff es)` work? Or it just doesn't fit with what your need to do?

**Reply by newhoggy** [reply]
Ideally, I should be able to use `hedgehog` and `effectful` functions together without lifting.

**Reply by newhoggy** [reply]
I'm now interested in implementing your suggestion, but I'm not sure where to start.

**Reply by newhoggy** [reply]
I have this:

```
type TestEffects2 = '[IOE]

property :: ()
  => H.PropertyT (Eff TestEffects2) ()
  -> H.Property
property f =
  H.Property H.defaultConfig $
    withFrozenCallStack (H.evalM (hoist runEff f))
```

**Comment by newhoggy** [comment]
BTW, I've managed to do this with `polysemy`: https://github.com/haskell-works/hw-polysemy/blob/main/test/HaskellWorks/Polysemy/HedgehogSpec.hs

**Comment by newhoggy** [comment]
I was able to get `effectful` to worth with `TestT IO`: https://github.com/haskell-works/effectful-zoo

Thanks for the suggestion @ocharles !

**Reply by arybczak** [reply]
Nice 👍 one suggestion: please don't put INLINE pragmas on functions unless you've benchmarked and concluded they help. GHC inlining heuristics are pretty good, so most of the time these pragmas just lead to longer compile times and code bloat.

**Comment by newhoggy** [comment]
I'm trying to implement `EvalM` to call `evalM` from `hedgehog` and have trouble aligning the types:

https://github.com/haskell-works/effectful-zoo/blob/main/components/hedgehog/Effectful/Zoo/Hedgehog/Effect/Hedgehog.hs#L21

The case that needs to be implemented is here:

https://github.com/haskell-works/effectful-zoo/blob/e2d2087c2c69a11d9b1d3d5d0f4e8dd6f9be9bc1/components/hedgehog/Effectful/Zoo/Hedgehog/Run.hs#L56

I'm running into difficulty because `evalM` takes a callback that is not `IO`:

https://hackage.haskell.org/package/hedgehog-1.5/docs/src/Hedgehog.Internal.Property.html#evalM

What should I do here?

I'm okay if it calls `evalIO` instead.

**Reply by newhoggy** [reply]
I moved my question to its own Q&A: https://github.com/haskell-effectful/effectful/discussions/273

**Comment by newhoggy** [comment]
I figured out a way to support hedgehog property tests with effectful:

https://github.com/haskell-works/effectful-zoo/pull/12
